tests/integration/backward_compatible/sql_test.py

```diff
@@ -0,0 +1,590 @@
+        self._populate_map()
+        result = self.client.sql.execute("SELECT this FROM %s" % self.map_name)
+
+        six.assertCountEqual(
+            self, [i for i in range(10)], [row.get_object_with_index(0) for row in result]
+        )
+
+        with self.assertRaises(ValueError):
+            for _ in result:
+                pass
+
+    def test_blocking_iterator_with_multi_paged_result(self):
+        self._populate_map()
+        statement = SqlStatement("SELECT __key FROM %s" % self.map_name)
+        statement.cursor_buffer_size = 1  # Each page will contain just 1 result
+        result = self.client.sql.execute_statement(statement)
+
+        six.assertCountEqual(
+            self, [i for i in range(10)], [row.get_object_with_index(0) for row in result]
+        )
+
+    def test_iterator(self):
+        self._populate_map()
+        result = self.client.sql.execute("SELECT __key FROM %s" % self.map_name)
+
+        iterator_future = result.iterator()
+
+        rows = []
+
+        def cb(f):
+            iterator = f.result()
+
+            def iterate(row_future):
+                try:
+                    row = row_future.result()
+                    rows.append(row.get_object_with_index(0))
+                    next(iterator).add_done_callback(iterate)
+                except StopIteration:
+                    pass
+
+            next(iterator).add_done_callback(iterate)
+
+        iterator_future.add_done_callback(cb)
+
+        def assertion():
+            six.assertCountEqual(
+                self,
+                [i for i in range(10)],
+                rows,
+            )
+
+        self.assertTrueEventually(assertion)
+
+    def test_iterator_when_iterator_requested_more_than_once(self):
+        self._populate_map()
+        result = self.client.sql.execute("SELECT this FROM %s" % self.map_name)
+
+        iterator = result.iterator().result()
+
+        rows = []
+        for row_future in iterator:
+            try:
+                row = row_future.result()
+                rows.append(row.get_object("this"))
+            except StopIteration:
+                break
+
+        six.assertCountEqual(self, [i for i in range(10)], rows)
+
+        with self.assertRaises(ValueError):
+            result.iterator().result()
+
+    def test_iterator_with_multi_paged_result(self):
+        self._populate_map()
+        statement = SqlStatement("SELECT __key FROM %s" % self.map_name)
+        statement.cursor_buffer_size = 1  # Each page will contain just 1 result
+        result = self.client.sql.execute_statement(statement)
+
+        iterator = result.iterator().result()
+
+        rows = []
+        for row_future in iterator:
+            try:
+                row = row_future.result()
+                rows.append(row.get_object_with_index(0))
+            except StopIteration:
+                break
+
+        six.assertCountEqual(self, [i for i in range(10)], rows)
```

Comment:
I don't think we need `cursor_buffer_size = 1` versions of several tests. Just one of cursor_buffer_size test to test pagination is enough.

Reply:
Different iterators might behave differently. So, it makes sense to test them with the same set of inputs